File: typescript/packages/core/src/commands/builtin/generic_bind/builders/touch.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { IOResult } from '../../../../io/types.ts'
import { fsStrerror, isFsError } from '../../../../utils/errors.ts'
import { specOf } from '../../../spec/builtins.ts'
import { FlagView } from '../../../spec/flag_view.ts'
import { type Builder, requireOp, resolveGlobOf } from '../adapter.ts'

const ENC = new TextEncoder()

export const TOUCH_BUILDER: Builder = {
  name: 'touch',
  write: true,
  fn: async (ops, accessor, paths, _texts, opts) => {
    if (paths.length === 0) {
      return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('touch: missing operand\n') })]
    }
    const idx = opts.index ?? undefined
    const write = requireOp(ops.write, 'write')
    const exists = requireOp(ops.exists, 'exists')
    const resolved = await resolveGlobOf(ops)(accessor, paths, idx)
    const createOnly = new FlagView(opts.flags, specOf('touch')).asBool('c')
    const writes: Record<string, Uint8Array> = {}
    const errors: string[] = []
    for (const p of resolved) {
      if (createOnly) continue
      if (await exists(accessor, p)) continue
      try {
        await write(accessor, p, new Uint8Array(0))
      } catch (err) {
        // One unusable operand is not an aborted command: GNU reports it
        // and still touches the remaining ones.
        if (!isFsError(err)) throw err
        errors.push(`touch: cannot touch '${p.virtual}': ${String(fsStrerror(err))}`)
        continue
      }
      writes[p.mountPath] = new Uint8Array(0)
    }
    const stderr = errors.length > 0 ? ENC.encode(errors.join('\n') + '\n') : null
    return [null, new IOResult({ writes, stderr, exitCode: errors.length > 0 ? 1 : 0 })]
  },
}

File: python/mirage/commands/builtin/generic_bind/builders/touch.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.base import Accessor
from mirage.commands.builtin.generic_bind.adapter import (Builder, CommandIO,
                                                          Operation)
from mirage.commands.config import CommandOpts
from mirage.commands.spec import SPECS
from mirage.commands.spec.flag_view import FlagView
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec
from mirage.utils.errors import FS_ERRORS, fs_strerror


async def touch(ops: CommandIO, accessor: Accessor, paths: list[PathSpec],
                texts: list[str],
                opts: CommandOpts) -> tuple[ByteSource | None, IOResult]:
    c = FlagView(opts.flags, spec=SPECS["touch"]).as_bool("c")
    if not ops.is_mounted(accessor) or not paths:
        raise ValueError("touch: missing operand")
    paths = await ops.resolve_glob(accessor, paths, opts.index)
    exists = ops.require(Operation.EXISTS)
    write = ops.require(Operation.WRITE)
    created: dict[str, ByteSource] = {}
    errors: list[str] = []
    for p in paths:
        if c:
            continue
        if await exists(accessor, p):
            continue
        try:
            await write(accessor, p, b"")
        except FS_ERRORS as exc:
            # One unusable operand is not an aborted command: GNU reports
            # it and still touches the remaining ones.
            errors.append(f"touch: cannot touch '{p.virtual}': "
                          f"{fs_strerror(exc)}")
            continue
        created[p.mount_path] = b""
    stderr = ("\n".join(errors) + "\n").encode() if errors else None
    return None, IOResult(writes=created,
                          stderr=stderr,
                          exit_code=1 if errors else 0)


BUILDER = Builder('touch', touch, write=True)
